doop.c

```diff
@@ -657,16 +657,68 @@ Magic and tainting are handled.
 =cut
 */
 
+PERL_STATIC_INLINE SV *
+S_do_join_inner(pTHX_ SV *lhs, SV *rhs)
+{
+    PERL_ARGS_ASSERT_DO_JOIN_INNER;
+
+    /* lhs has already had SvGETMAGIC called */
+    if (SvGAMAGIC(rhs)) {
+        SvGETMAGIC(rhs);
+        rhs = sv_mortalcopy_flags(rhs, 0);
+    }
+
+    int applies_flags = 0;
+    /* LD_LIBRARY_PATH=`pwd` perf stat ./perl -e '
+           my @x = ("a" x 100) x 1000;
+           for my $i (0 .. 10000) { my $z = join(",", @x); }'
+     * We speed this up by 20% (about) with a check for both ROK and RMG,
+     * because those must be set if an SV has any overload. */
+    U32 flags = SvFLAGS(lhs)|SvFLAGS(rhs);
+    if (UNLIKELY((flags & (SVf_ROK|SVs_GMG))) &&
+        UNLIKELY(amagic_applies(lhs, concat_amg, applies_flags)
```

Comment:
and then replace this call to amagic_applies() with lhs_has_mg.

This will halve the number of times this code calls amagic_applies() for a given string.